### ISSUE TITLE
New rule: character-whitelist to limit characters allowed

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#10747]
**Please describe what the rule should do:**

The _character-whitelist_ rule checks that code source contains only ASCII characters. It helps also to detect [unicode homoglyph](https://en.wikipedia.org/wiki/Homoglyph#Unicode_homoglyphs) problems.

This rule has two options:

- A string listing all characters allowed. The default value is the list of [95 printable ASCII characters](https://en.wikipedia.org/wiki/ASCII#Printable_characters), horizontal tab (`\t`), line feed (`\n`) and carriage return (`\r`).
- An object to configure characters allowed in different parts of the code: `"strings"`, `"comments"`, `"regexps"`, and `"templates"`. The values can be:
  - `null` (default): doesn't check characters for this part;
  - a string of characters which complete global list of characters allowed.

Here is an example configuration:

```JSON
{
    "character-whitelist": ["error", {
        "strings": "",
        "comments": "ÀÂÄÇÉÈÊËÎÏÔÖÙÛÜŸÆŒàâäçéèêëîïôöùûüÿæœ",
        "regexps": "",
        "templates": "",
    }]
}
```

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
console.log(/[a–c]/.test("b"));
// "–" is U+2013 character (en dash).
```

```js
const baz = "quz";
console.log(bаz);
// "а" (of bаz in console.log) is U+0430 character (a cyrillic).
// ESLint returns "'bаz' is not defined." but it doesn't help user to find
// the error.
```

```js
console.log("quux" == "qսսx");
// "սս" (of second qսսx) is U+057D (u armenian).
```

```js
console.log(`corge` == `cοrge`);
// "ο" (of second cοrge) is U+03BF (omicron).
```

```js
// FIХME Urgent !!!
// "Х" is U+0425 (kha cyrillic).
```
**Why should this rule be included in ESLint (instead of a plugin)?**

It's a good practice to use only ASCII characters in code source. This rule is similar to [no-irregular-whitespace](https://eslint.org/docs/rules/no-irregular-whitespace) rule but more general. It can replace [no-tabs](https://eslint.org/docs/rules/no-tabs).

**Comment by lll000111** [comment]
Here are some stories showing why this can be very useful: https://old.reddit.com/r/programming/comments/98crp8/how_to_write_unmaintainable_code/e4f5oop/
